### PR TITLE
a modal route example with next js

### PR DESCRIPTION
**Pull request by mohammad-amin-hesam**
as i searched in google some ppl want to know how they can make a route based modal like react-router-dom : https://reacttraining.com/react-router/web/example/modal-gallery

this is a modal route example.
I hope it is useful :)

### CODE REVIEW

**Comment by BjoernRave** [comment]
wouldn't it be easier to set a queryparam on the page you want to open the modal, check for that queryparam and maybe open a modal then?

File: examples/next-modal-route/pages/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import Modal from "../components/Modal";
+
+const App = () => {
+  const [modal, setModal] = useState(false);
+
+  const _modal = useRef(null);
+
+  const handleClick = () => {
+    setModal(true);
+    window.history.pushState("", "", "/test");
```
Comment (timneutkens) [inline examples/next-modal-route/pages/index.js]: This breaks Next.js / back button behavior, it should use the Next.js router instead.

**Comment by singlepageprogrammer** [comment]
if i use next js routing it's illogical bc in default ```window.history``` changes url without change component and next js or any other router libraries used for changing component after push history ...
now u think we say to next js router : hey plz don't change component i want to open a modal.
and next js will answer to you : that's why u using me while in default window don't change your component ?

i think using window.history.pushState is the best way for handle that

**Comment by timneutkens** [comment]
You'll want to add `shallow: true`

**Comment by timneutkens** [comment]
Going to close this to prevent stale open PRs, but we'd be happy to merge an example like this 👍

**Comment by joonhocho** [comment]
@timneutkens Can we reopen this? This is useful PR.
There is a problem using `shallow: true` because it only works when route is already on the same page.
For example, if user is at `/posts/list` and wants to show post modally with url `/post/postId`, then shallow routing does not work. It still navigates to the post page because the route is different.

**Comment by lkbr** [comment]
I've asked a similar question here: https://spectrum.chat/next-js/general/what-is-the-best-way-to-create-the-instagram-photo-modal-routing-experience-on-dynamic-pages-with-next-9~5695efed-d505-446b-8474-1ce2117e19d0

Not sure how to do this type of modal the Next.js ordained way.

**Comment by FarazPatankar** [comment]
+1 to reopening this. Facing the same issue and see questions on here and on Spectrum unanswered. What is the best approach to doing this if the route isn't on the same page or should we just avoid doing it if we're using Next?

**Pull request closed without merging**